src/Symfony/Component/DomCrawler/DomCrawler.php

```diff
@@ -0,0 +1,350 @@
+            }
+        } else {
+            $dom = new \DOMDocument('1.0', $charset);
+            $dom->validateOnParse = true;
+
+            if ('' !== trim($content)) {
+                @$dom->loadXML($content, $options);
+            }
+        }
+
+        libxml_use_internal_errors($internalErrors);
+
+        $this->addDocument($dom);
+
+        $this->isHtml = false;
+    }
+
+    /**
+     * Adds a \DOMDocument or a \DOM\Document to the list of nodes.
+     */
+    public function addDocument(\DOMDocument|\DOM\Document $dom): void
+    {
+        if ($dom->documentElement) {
+            $this->addNode($dom->documentElement);
+        }
+    }
+
+    /**
+     * Adds a \DOMNodeList or a \DOM\NodeList to the list of nodes.
+     */
+    public function addNodeList(\DOMNodeList|\DOM\NodeList $nodes): void
+    {
+        foreach ($nodes as $node) {
+            if ($node instanceof \DOMNode || $node instanceof \DOM\Node) {
+                $this->addNode($node);
+            }
+        }
+    }
+
+    /**
+     * Adds a \DOMNode or a \DOM\Node instance to the list of nodes.
+     */
+    public function addNode(\DOMNode|\DOM\Node $node): void
+    {
+        if ($node instanceof \DOMDocument || $node instanceof \DOM\Document) {
+            $node = $node->documentElement;
+        }
+
+        if (null !== $this->domDocument && $this->domDocument !== $node->ownerDocument) {
+            throw new \InvalidArgumentException('Attaching DOM nodes from multiple documents in the same crawler is forbidden.');
+        }
+
+        $this->domDocument ??= $node->ownerDocument;
+        if ($node->ownerDocument instanceof \DOMDocument) {
+            $this->document = $node->ownerDocument;
+        }
+
+        // Don't add duplicate nodes in the Crawler
+        if (\in_array($node, $this->nodes, true)) {
+            return;
+        }
+
+        $this->nodes[] = $node;
+    }
+
+    protected function sibling(\DOMNode|\DOM\Node $node, string $siblingDir = 'nextSibling'): array
+    {
+        $nodes = [];
+
+        $currentNode = $this->getDomNode(0);
+        do {
+            if ($node !== $currentNode && \XML_ELEMENT_NODE === $node->nodeType) {
+                $nodes[] = $node;
+            }
+        } while ($node = $node->$siblingDir);
+
+        return $nodes;
+    }
+
+    private function parseHtml5(string $htmlContent, string $charset = 'UTF-8'): \DOMDocument|\DOM\HTMLDocument
+    {
+        $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
+
+        if ($this->options & self::CRAWLER_USE_NATIVE_PARSER) {
+            return \DOM\HTMLDocument::createFromString($htmlContent, \DOM\HTML_NO_DEFAULT_NS);
+        }
+
+        return $this->html5Parser->parse($htmlContent);
+    }
+
+    private function parseXhtml(string $htmlContent, string $charset = 'UTF-8'): \DOMDocument|\DOM\XMLDocument
+    {
+        $htmlContent = $this->convertToHtmlEntities($htmlContent, $charset);
```

Comment:
is this needed when parsing with the new extension or does it support UTF-8 properly ?

Reply:
also, do _HTML_ entities make sense when parsing _XML_ documents ?

Reply:
I'd say yes, even if the name seems weird in this case, the conversion to HTML entities is also required. Otherwise, many tests break. Not sure if HTML entities have another naming that could better fit 🤔